src/transformation/mesh_intersection/mesh_intersection.rs

```diff
@@ -256,393 +348,551 @@ fn extract_connected_components(
-                    for inter in intersections {
-                        let f1 = convert_fid(mesh1, i1, inter.f1);
-                        let f2 = convert_fid(mesh2, i2, inter.f2);
-                        let orig_fid = [inter.f1, inter.f2];
-                        let key = (f1, f2);
-                        let ins = *intersection_points
-                            .entry(key)
-                            .or_insert([inter.p1, inter.p2]);
-
-                        let _ = insert_point(ins, key, orig_fid, 0);
-                        let _ = insert_point(ins, key, orig_fid, 1);
-                    }
-                }
+fn insert_into_set(
+    position: Point3<Real>,
+    point_set: &mut RTree<TreePoint>,
+    epsilon: Real,
+) -> usize {
+    let point_count = point_set.size();
+    let point_to_insert = TreePoint {
+        point: position,
+        id: point_count,
+    };
+
+    match point_set.nearest_neighbor(&point_to_insert) {
+        Some(tree_point) => {
+            if (tree_point.point - position).norm_squared() <= epsilon {
+                tree_point.id
+            } else {
+                point_set.insert(point_to_insert);
+                debug_assert!(point_set.size() == point_count + 1);
+                point_count
             }
         }
+        None => {
+            point_set.insert(point_to_insert);
+            debug_assert!(point_set.size() == point_count + 1);
+            point_count
+        }
     }
-
-    extract_result(
-        pos12,
-        mesh1,
-        flip1,
-        mesh2,
-        flip2,
-        &spade_handle_to_intersection,
-        &intersection_points,
-        &triangulations1,
-        &triangulations2,
-        new_vertices12,
-        new_indices12,
-    );
 }
 
-fn convert_fid(mesh: &TriMesh, tri: u32, fid: FeatureId) -> FeatureId {
-    match fid {
-        FeatureId::Edge(eid) => {
-            let topology = mesh.topology().unwrap();
-            let half_edge_id = topology.face_half_edges_ids(tri)[eid as usize];
-            let half_edge = &topology.half_edges[half_edge_id as usize];
-            // NOTE: if the twin doesn’t exist, it’s equal to u32::MAX. So the `min` will
-            //       automatically filter it out.
-            FeatureId::Edge(half_edge_id.min(half_edge.twin))
+fn smallest_angle(points: &[Point3<Real>]) -> Real {
+    let n = points.len();
+
+    let mut worst_cos = 2.0;
+    for i in 0..points.len() {
+        let d1 = (points[i].coords - points[(i + 1) % n].coords).normalize();
+        let d2 = (points[(i + 2) % n].coords - points[(i + 1) % n].coords).normalize();
+
+        let cos = d1.dot(&d2);
+
+        if cos < worst_cos {
+            worst_cos = cos.abs();
         }
-        FeatureId::Vertex(vid) => FeatureId::Vertex(mesh.indices()[tri as usize][vid as usize]),
-        FeatureId::Face(_) => FeatureId::Face(tri),
-        FeatureId::Unknown => FeatureId::Unknown,
     }
+
+    worst_cos.acos() * 180. / PI as Real
```

Comment:
Looks like my previous review’s comment about sticking to radians went unnoticed.